### DISCUSSION TITLE
v2.0

### DISCUSSION THREAD

**Discussion by soghomon-b** [bojai-org/bojai#6] [Announcements]
<html>
<body>
<html><head></head><body><h1>Bojai 2.0&nbsp;– Release Notes (August&nbsp;1,&nbsp;2025)</h1>
<p><strong>Audience:</strong> ML researchers, engineers, and technical users<br>

<hr>
<h2>🚀 Overview</h2>
<p>Bojai 2.0 introduces major architectural and usability upgrades, grounded in modularity, observability, and user customization. Highlights include:</p>
<ul>
<li>
<p>A rewritten <strong>How‑To Custom Pipeline Guide</strong>, complete with a five‑module boilerplate structure.</p>
</li>
<li>
<p>An <strong>LLM‑driven “Data Processing Agent”</strong> (beta, Ollama‑based) for rapid dataset ingestion.</p>
</li>
<li>
<p><strong><code inline="">modify</code> / <code inline="">checkout</code> CLI(s)</strong> for in‑place editing and auditing of built‑in pipelines.</p>
</li>
<li>
<p>A built‑in <strong><code inline="">Visualizer</code> + <code inline="">Logger</code></strong> framework for epoch‑level metrics monitoring.</p>
</li>
<li>
<p>A <strong><code inline="">save</code>/<code inline="">load</code> snapshot system</strong> to persist and resume training mid‑workflow.</p>
</li>
</ul>
<p>Detailed docs are available in the “How‑to guides” and “References” sections of our documentation.</p>
<hr>
<h2>📘 New / Improved User‑Facing Components</h2>
<h3>1. <strong>Improved “How to Code a Custom Pipeline” guide</strong></h3>
<p>The new walkthrough (updated July&nbsp;19,&nbsp;2025) covers all five Python modules required for a custom Bojai pipeline:</p>
<pre><code class="language-text">custom_data_processor.py  
custom_pipeline_user.py  
custom_trainer.py  
global_vars.py  
model.py
</code></pre>
<p>Rather than abstracting these away, this guide walks through each file in depth—including methods for preprocessing, training loops, eval logic, and configuration parameters—so you can fully control the pipeline stages. This replaces the more limited custom pipeline doc from v1.0.(<a href="https://bojai-documentation.web.app/how-to-guides/how-to-use-the-agent/?utm_source=chatgpt.com" title="How to use the data processing agent">bojai-documentation.web.app</a>)</p>
<hr>
<h3>2. <strong>LLM‑powered Data Processing Agent (beta)</strong></h3>
<p>Leveraging a local Ollama installation, the new agent can bootstrap your <code inline="">get_inputs_outputs()</code> method automatically based on natural‑language prompts during the CLI initialization flow.<br>
<strong>CLI flow sample:</strong></p>
<pre><code class="language-text">Would you like to use the agent? [Y/n] → Y  
Does your data contain images? [Y/n] → n  
Describe dataset layout: "text file with input and output comma‑separated"  
</code></pre>
<p>Under the hood, it iteratively (max&nbsp;3 attempts) generates, injects, and unit‑tests a method to parse data, then replaces <code inline="">custom_data_processor.py</code> upon success. Ollama must be pre‑installed, and the agent retries automatically if the function fails initially.(<a href="https://bojai-documentation.web.app/how-to-guides/how-to-use-the-agent/?utm_source=chatgpt.com" title="How to use the data processing agent">bojai-documentation.web.app</a>)<br>
<strong>Note:</strong> Still in beta; working offline only with a local LLM.</p>
<hr>
<h3>3. <strong>Modify &amp; Checkout pipelines (new CLI commands)</strong></h3>
<ul>
<li>
<p><strong><code inline="">bojai modify &lt;pipeline-id&gt;</code></strong> creates a clone of any built‑in pipeline in your workspace, exposing all underlying files.</p>
</li>
<li>
<p><strong><code inline="">bojai checkout &lt;pipeline-id&gt;</code></strong> launches a CLI/UI editor on the cloned workspace for easy inspection or editing.<br>
This grants full access to tweak data loaders, model architecture, training logic, or hyper‑parameters without breaking registry support or existing tooling. It’s key for researchers and auditors who need transparency or versioned pipeline snapshots.</p>
</li>
</ul>
<hr>
<h3>4. <strong>Built‑in Logging + Visualizer</strong></h3>
<p>A <code inline="">Logger</code> class now integrates deeply into <code inline="">trainer.py</code>. Example usage:</p>
<pre><code class="language-python">self.logger.log(epoch=epoch,
                loss=loss,
                train=training_score,
                valid=validation_score)
And for final eval:
self.logger.log(eval_score=final_score)
</code></pre>
<p>Later, the <code inline="">Visualizer</code> module reads these logs and plots:</p>
<ul>
<li>
<p><strong>Loss vs epoch</strong></p>
</li>
<li>
<p><strong>Training vs validation accuracy</strong></p>
</li>
<li>
<p><strong>Final evaluation scores</strong></p>
</li>
</ul>
<p>All views are paired by epoch and rendered in CLI or web UI. This solves post‑hoc visibility problems and offers full traceability across runs via the visualizer reference documentation.</p>
<hr>
<h3>5. <strong>Save and Load Training Session State</strong></h3>
<p>You can now <strong>persist a training session</strong> mid‑run—including model parameters, optimizer state, epoch counter, loss history, and evaluation scores—and <strong>reload it by name</strong> at a later time.</p>
<ul>
<li>
<p><strong><code inline="">bojai save &lt;session-name&gt;</code></strong> serializes the full state.</p>
</li>
<li>
<p><strong><code inline="">bojai load &lt;session-name&gt;</code></strong> reinstates it for resuming (e.g. after interruption, or switching branches).</p>
</li>
</ul>
<p>This is useful for long‑running experiments, hyperparameter tuning, or shared workflows in multi‑user settings.</p>
<h3>📌 Behavior note:</h3>
<ul>
<li>
<p>Checkpoint files are versioned in under <code inline="">bojai_sessions/&lt;user&gt;/&lt;session-name&gt;</code>.</p>
</li>
<li>
<p>On load, Bojai verifies file integrity before proceeding.</p>
</li>
<li>
<p>Resuming wraps epochs so training continuity (epoch&nbsp;N+1 → N+2) is preserved.</p>
</li>
</ul>
<hr>

<hr /><em>This discussion was created from the release <a href='https://github.com/bojai-org/bojai/releases/tag/v2.0'>v2.0</a>.</em>
